File: Cpp/0983_minimum_cost_for_tickets.cpp
// https://leetcode.cn/problems/minimum-cost-for-tickets/

class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        vector<int> dp(days.back()+1, 0);
        unordered_set<int> s(days.begin(), days.end());
        for (int i = 1; i < days.back()+1; ++i) {
            if (s.find(i) == s.end()) {
                dp[i] = dp[i-1];
            }
            else {
                int tmp1 = dp[i-1] + costs[0];
                int tmp2 = dp[max(i-7, 0)] + costs[1];
                int tmp3 = dp[max(i-30, 0)] + costs[2];
                dp[i] = min(tmp1, min(tmp2, tmp3));
            }
        }
        return dp.back();
    }
};

File: Python3/0983_minimum_cost_for_tickets.py
# https://leetcode.cn/problems/minimum-cost-for-tickets/


class Solution:
    def mincostTickets(self, days, costs):
        """
        Inputs:
            days, costs: list[int]
        Outputs:
            res: int
        """
        dp = [0 for _ in range(days[-1]+1)]
        s = set(days)
        for i in range(1, days[-1]+1):
            if i not in s:
                dp[i] = dp[i-1]
            else:
                dp[i] = min(dp[i-1] + costs[0],
                            dp[max(i-7, 0)] + costs[1],
                            dp[max(i-30, 0)] + costs[2])
        return dp[-1]